icu4c/source/i18n/simpletz.cpp

```diff
@@ -519,10 +519,13 @@ SimpleTimeZone::getOffsetFromLocal(UDate date, UTimeZoneLocalOption nonExistingT
 
     rawOffsetGMT = getRawOffset();
     int32_t year, month, dom, dow;
+    // For an integer n and representable ⌊x/n⌋, ⌊RN(x/n)⌋=⌊x/n⌋, where RN is
+    // rounding to nearest.
     double day = uprv_floor(date / U_MILLIS_PER_DAY);
-    // "(int32_t) (x + y)" will produce a 1-off rounding error while
-    // y > 1e8 and x is beteen (0, -1e-9).
-    // Fix it by wrapping x with uprv_flooring()
+    // For doubles x and n, where n is an integer and ⌊x+n⌋ < 2³¹, the
+    // expression `(int32_t) (x + n)` evaluated with rounding to nearest
+    // differs from ⌊x+n⌋ if 0 < ⌈x⌉−x ≪ x+n, as `x + n` is rounded up to
+    // n+⌈x⌉ = ⌊x+n⌋ + 1.  Rewriting it as ⌊x⌋+n makes the addition exact.
```

Comment:
Two comments:
1) Maybe it's because I'm still trying to wake up this morning, or maybe it's because I'm a music major, but I can't make head nor tail out of either of these comments.  I think they either need to be made more general, go into a lot more detail, or point to an external source (some general explanation of IEEE floating-point math?) that might explain better what's wrong and how this fixes it.
2) It doesn't help that these same indecipherable descriptions appear in four different places in our code base.  Maybe it's worth moving the actual math into a utility library somewhere-- not because the code is that complicated, but because the explanation is.

Reply:
I definitely agree with 2, and we could then address 1 in the « go into more detail » direction, rather than handwaving a ≪ while desperately trying to squeeze it all in four lines. I guess `floorDivide` is already trying to be that general-purpose utility?

Reply:
yes, I think ClockMath::floorDivide should the place. I really dislike it take the 3rd arg as reference and modify the ref value. I change it to a pointer here so in the caller call it is obvious that value could be changed after the call.